apollo-router/src/spec/query.rs

```diff
@@ -593,6 +593,31 @@ impl Query {
                         continue;
                     }
 
+                    if name.as_str() == TYPENAME {
+                        let input_value = input
+                            .get(field_name.as_str())
+                            .cloned()
+                            .filter(|v| v.is_string())
+                            .unwrap_or_else(|| {
+                                Value::String(ByteString::from(
+                                    parent_type.inner_named_type().as_str().to_owned(),
+                                ))
+                            });
```

Comment:
this is specifically a check to replace a null `__typename` with the type we currently know. Maybe in general, we should make sure to replace the typename with the one we know from the query and the API schema